rfcs/45-stageless.md

```diff
@@ -0,0 +1,761 @@
+
+### Storage
+
+```rust
+struct SystemRegistry {
+    // Systems can only be run on the `World` that initialized them.
+    world_id: Option<WorldId>,
+    // Maps a labels to identifiers.
+    // We need to be able to quickly look up specific systems and sets or iterate over them.
+    // `RegId` is unique identifier, generated on insertion.
+    ids: HashMap<BoxedSystemLabel, RegId>,
+    // ID of next thing to be registered.
+    next_id: RegId,
+    // Permanent system storage.
+    systems: HashMap<RegId, Option<BoxedSystem>>,
+    // Permanent conditions storage.
+    conditions: HashMap<RegId, Option<Vec<BoxedRunCondition>>>,
+    // Temporary storage and cached metadata for executing registered sets.
+    schedules: HashMap<RegId, Option<Schedule>>,
+    // Graph-related metadata for sets. Used for run-time modification.
+    set_metadata: HashMap<RegId, SetMetadata>,
+    // The "global" hierarchy of everything in the registry. Used for run-time modification.
+    // `Graph` is a bikeshed-avoidance graph storage structure.
+    hierarchy: Graph<RegId>,
+    // Stuff that hasn't been initialized yet.
+    uninit_nodes: Vec<(RegId, Config)>,
+}
+
+enum RegId {
+    System(u64),
+    Set(u64),
+}
+
+struct SetMetadata {
+    // The graph of the systems and sets directly under this set.
+    graph: Graph<RegId>,
+    // The cached topological order for `graph`.
+    topsort: Vec<RegId>,
+    flattened: Graph<RegId>,
+    flattened_topsort: Vec<RegId>,
+    hierarchy: Graph<RegId>,
+    hierarchy_topsort: Vec<RegId>,
+    // The combined component access of everything under this set.
+    combined_access: FilteredAccessSet<ComponentId>,
+    // True if the schedule for this set needs to be regenerated.
+    modified: bool,
+}
+```
+
+### Run conditions (Run criteria)
+
+Conditions are systems with several key constraints:
+
+- They must return a `bool`.
+- They cannot have system params that grant mutable access to data.
+  - We don't have order to them.
+  - Avoids side effects.
+- They cannot capture or store local data.
+  - Avoid footgun: local data would be modified even if another condition stops the guarded system from running.
+
+When the executor is presented with a ready system, we do the following:
+
+- Check that all the data accessed by the system, its run conditions (RC), and the RC of any sets it is under (that haven't been evaluated yet) is available.
+  - If not, leave the system in the ready queue and move onto the next one.
+- If yes, evaluate the RC of the non-evaluated sets that the system is under in hierarchical order (from the outermost set to the innermost set).
+  - If any set's RC return false, mark all the systems (incl. current one) and sets under that set as completed/evaluated.
+  - If any set's RC return true, mark the set as evaluated.

```

Comment:
Shouldn't it be `If all set's RC return true...`? I mean, a given set can have multiple RC's, so all should return true, before the set be evaluated, right?

Maybe just a typo.